From 68f64322a1b220c325dfa10afaf3041e83def3ef Mon Sep 17 00:00:00 2001
From: Jani Hautakangas <jani@igalia.com>
Date: Tue, 24 Sep 2024 03:16:13 -0700
Subject: [PATCH 10/12] [GTK][WPE] IPC::Connection fails to unset FD_CLOEXEC
 https://bugs.webkit.org/show_bug.cgi?id=280256

Reviewed by Carlos Garcia Campos.

There's a logic error in UniStdExtrasUnix.cpp where unsetCloseOnExec(int)
expects fnctl(...) F_GETFD to return -1 as a flags value before trying to
remove FD_CLOEXEC flag. However, -1 indicates error so FD_CLOEXEC never gets
removed and calling function in ConnectionUnix createPlatformConnection has
RELEASE_ASSERT which leads to crash.

unsetCloseOnExec should check if fnctl(...) F_GETFD return value is
!= -1 before setting new flag value.

* Source/WTF/wtf/unix/UniStdExtrasUnix.cpp:
(WTF::unsetCloseOnExec):

Canonical link: https://commits.webkit.org/284153@main
---
 Source/WTF/wtf/unix/UniStdExtrasUnix.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Source/WTF/wtf/unix/UniStdExtrasUnix.cpp b/Source/WTF/wtf/unix/UniStdExtrasUnix.cpp
index b9a63fbc76a5..c45c75d96f0e 100644
--- a/Source/WTF/wtf/unix/UniStdExtrasUnix.cpp
+++ b/Source/WTF/wtf/unix/UniStdExtrasUnix.cpp
@@ -47,7 +47,7 @@ bool unsetCloseOnExec(int fileDescriptor)
     int returnValue = -1;
     do {
         int flags = fcntl(fileDescriptor, F_GETFD);
-        if (flags == -1)
+        if (flags != -1)
             returnValue = fcntl(fileDescriptor, F_SETFD, flags & ~FD_CLOEXEC);
     } while (returnValue == -1 && errno == EINTR);
 
-- 
2.43.0

